### ISSUE TITLE
Add Block: Calendar

### ISSUE THREAD

**Issue by melchoyce** [WordPress/gutenberg#1462]
Similar to: #1799, #1800, #1801
PRs where similar blocks are developed:  #7966, #7875, #4501 

## Attributes

- None, Left, Right, Center, Full Width?
- Display full-width

## States

Selected:

![calendar](https://user-images.githubusercontent.com/2846578/27554556-cca52de6-5a7c-11e7-8606-01b62007f286.png)

![calendar block selected](https://user-images.githubusercontent.com/2846578/27554568-d288284e-5a7c-11e7-84df-4d78d385b957.png)

Neutral (Full-Width):

![calendar block unselected](https://user-images.githubusercontent.com/2846578/27554569-d28e294c-5a7c-11e7-94f6-c887924de65a.png)

-----

Questions:
- Should we include alignment?
- If yes, should alignment be a default setting in the Block Level Formatting bar?
- Are there any other settings that should be included?

Related to #1011.

**Comment by nylen** [comment]
This seems like it would be more useful with a way to specify a list of events.  A few possible ways that data could be specified:

- Filter by post type, tag, category, etc.
- Manual entry
- Google calendar import
- ICS file import

**Comment by melchoyce** [comment]
@nylen To clarify, this is a conversion of the Calendar widget, which is a calendar of your posts.

**Comment by marsjaninzmarsa** [comment]
@nylen most use cases I can think about requires support for at least CPT.

**Comment by folletto** [comment]
> Should we include alignment?

If we support it not being full width, I think supporting alignment makes sense. The two go hand in hand for me.

I'd also ask if we should make "full width" the default.


> If yes, should alignment be a default setting in the Block Level Formatting bar?

Only if "full width" does.


> Are there any other settings that should be included?

Some come to my mind, but I don't think they make sense for the first version. Mentioning still as might be useful for the discussion:

1. Show a specific Month / Year
2. Show only a specific Category
3. Show only a specific Tag
4. Show only a specific Post / Custom Post Type (as suggested by @marsjaninzmarsa) 
5. _{{ Show only future posts (this is unsupported in WordPress, as future posts are automatically Scheduled instead of posted, but seems something that could build up to a nice way to feature future events for event sites) }}_

---

Also given the discussion right above, I might suggest to change "Calendar" to "Posts Calendar" or something more specific?

Which leads me to another consideration: is the "Calendar" just an alternative view of the Posts List Archive (#1464)?

**Comment by westonruter** [comment]
I believe this depends on having an endpoint for server-side rendering: #780

**Comment by karmatosed** [comment]
I think changing to 'Posts Calendar' makes sense. I've marked this as needing to get dev as I think it's ready for someone to dip in.

**Comment by karmatosed** [comment]
As the project moves towards merge proposal, we are moving issues that aren't needed for that to projects. This doesn't mean they won't get done, they totally can and that's why we're moving to projects. This allows us to focus on the issues that have to happen to Gutenberg. I am closing this but it will live in projects.

**Comment by melchoyce** [comment]
Took a step back and rethought what a calendar block could look like if we started from scratch. 

I focused on exploring the “what if’s” — what if you could schedule drafts from the Calendar block? What if you could reschedule posts? 

You can view what I have here: https://cloudup.com/c3HqgFiImw5

Or walk through it in this prototype: https://wp.invisionapp.com/share/VHHNY6CQZGS#/screens

cc @mtias @karmatosed

**Comment by melchoyce** [comment]
Thanks for your feedback, @afercia! We should definitely use whatever accessible calendar solution comes from #1311. 

> Ideally the available actions, in this case "Cancel" and "Select" should be placed after the content, otherwise users would be forced to navigate backwards to operate on them.

In this instance, do you think it's okay for the actions to visually appear at the top of the modal, as long as they're located correctly within the source order of the page?

**Comment by melchoyce** [comment]
Looking at standardizing these popups:

<img src="https://user-images.githubusercontent.com/2846578/39446780-a36c0f54-4c8d-11e8-8191-39070b2d531c.png" width="244" /><img src="https://user-images.githubusercontent.com/2846578/39446799-b8640150-4c8d-11e8-9eff-c78fdee07749.png" width="265" /><img src="https://user-images.githubusercontent.com/2846578/39446787-ad3aea64-4c8d-11e8-9604-6215e7370889.png" width="336" /><img src="https://user-images.githubusercontent.com/2846578/39446811-c0336fe2-4c8d-11e8-9db3-e33bb7d0ec97.png" width="242" />

**Comment by melchoyce** [comment]
Prototype updated: https://wp.invisionapp.com/share/VSHRRSCX47E#/293882456_Calendar-_Neutral

**Comment by iseulde** [comment]
Could this blocks be called "Post Calendar", to avoid confusion?

**Comment by gziolo** [comment]
@melchoyce, I have just checked the prototype you shared a long time ago. It looks quite complex for the first iteration. What would be the minimal version of Calendar of Posts? Should it be possible to switch between months or should we only render the current month?

**Comment by gziolo** [comment]
I have quickly enabled Calendar on my own website. This is what I see on the frontend:

![screen shot 2019-01-28 at 10 54 37](https://user-images.githubusercontent.com/699132/51828115-2f6f6800-22eb-11e9-8f43-6cca5932abbb.png)

Some observations:
- it renders the current months by default with a link to the previous month with a post published
- if I browser the post it sets the month for the calendar to align with the post, the same applies to archives
- it links months to archive pages for months
- it links days to archive pages for days

**Comment by melchoyce** [comment]
The v1 should more closely match the initial mockup, which is a 1:1 port of the existing calendar widget. Visually, I figured we could reuse the styles for the datepicker component.

**Comment by afercia** [comment]
Adding the Accessibility label in addition to the "needs..." one so the team can search for all the widget-blocks related issue in an easier way.

**Comment by gziolo** [comment]
@melchoyce I'm guessing you will want to open a new issue with some refinements planned for the future, right? :)

**Comment by melchoyce** [comment]
Yes, will do 👍
